### PR TITLE
[User management] Comment on time zone independent epoch extraction in Postgres and Oracle

### PR DESCRIPTION
**Pull request by pbatko-da**

### CODE REVIEW

File: ledger/participant-integration-api/src/main/resources/db/migration/oracle-appendonly/V8__add_user_managment.sql [hunk 1]
```diff
@@ -16,6 +16,18 @@ CREATE TABLE participant_user_rights (
...
+-- `to_timestamp('1-1-1970 00:00:00','MM-DD-YYYY HH24:Mi:SS')` returns a TIMESTAMP (i.e. without time zone)
+-- Oracle docs states:
+-- """
+-- Oracle performs all timestamp arithmetic in UTC time. For TIMESTAMP WITH LOCAL TIME ZONE, Oracle converts
+-- the datetime value from the database time zone to UTC and converts back to the database time zone after
+-- performing the arithmetic. For TIMESTAMP WITH TIME ZONE, the datetime value is always in UTC, so no conversion is necessary.
+-- """
+-- so it isn't explicit/clear about arithmetic on plain TIMESTAMP type.
+-- However empirical tests give evidence that TIMESTAMP arithmetic is done as if TIMESTAMP values were in UTC time.
+
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/oracle-appendonly/V8__add_user_managment.sql:30]: Empirical tests.
The values of diff (time difference computed with `timestamp`) and `diffz` (time differene computed with `timestamp with time zone`) should be the same.



```
CREATE VIEW tmp_view AS
SELECT
    sessiontimezone AS zn,
    cz,
    c,
    tz,
    t,
    cz - tz AS diffz,
    c - t AS diff
FROM (
    SELECT
        cz,
        sys_extract_utc(cz) c,
        t,
        tz
    FROM (
        SELECT
            current_timestamp AS cz,
            to_timestamp_tz('1-1-1970 00:00:00 0:00', 'MM-DD-YYYY HH24:Mi:SS TZH:TZM') AS tz,
            to_timestamp('1-1-1970 00:00:00','MM-DD-YYYY HH24:Mi:SS') AS t
        FROM dual
        )
    )
;
DESC tmp_view;

    Name    Null?                            Type
________ ________ _______________________________
ZN                VARCHAR2(75)
CZ                TIMESTAMP(6) WITH TIME ZONE
C                 TIMESTAMP(6)
TZ                TIMESTAMP(9) WITH TIME ZONE
T                 TIMESTAMP(9)
DIFFZ             INTERVAL DAY(9) TO SECOND(9)
DIFF              INTERVAL DAY(9) TO SECOND(9)


SQL> ALTER SESSION SET TIME_ZONE='Asia/Hong_Kong';

Session altered.

SQL> SELECT * from tmp_view;

               ZN                                            CZ                              C                                 TZ                              T                     DIFFZ                      DIFF
_________________ _____________________________________________ ______________________________ __________________________________ ______________________________ _________________________ _________________________
Asia/Hong_Kong    22/01/21 20:59:46,989146000 ASIA/HONG_KONG    22/01/21 12:59:46,989146000    70/01/01 00:00:00,000000000 GMT    70/01/01 00:00:00,000000000    +19013 12:59:46.989146    +19013 12:59:46.989146

SQL> ALTER SESSION SET TIME_ZONE='Europe/Warsaw';

Session altered.

SQL> SELECT * from tmp_view;

              ZN                                           CZ                              C                                 TZ                              T                     DIFFZ                      DIFF
________________ ____________________________________________ ______________________________ __________________________________ ______________________________ _________________________ _________________________
Europe/Warsaw    22/01/21 13:59:59,586891000 EUROPE/WARSAW    22/01/21 12:59:59,586891000    70/01/01 00:00:00,000000000 GMT    70/01/01 00:00:00,000000000    +19013 12:59:59.586891    +19013 12:59:59.586891

SQL> ALTER SESSION SET TIME_ZONE='-07:00';

Session altered.

SQL> SELECT * from tmp_view;

       ZN                                    CZ                              C                                 TZ                              T                     DIFFZ                      DIFF
_________ _____________________________________ ______________________________ __________________________________ ______________________________ _________________________ _________________________
-07:00    22/01/21 06:00:22,069809000 -07:00    22/01/21 13:00:22,069809000    70/01/01 00:00:00,000000000 GMT    70/01/01 00:00:00,000000000    +19013 13:00:22.069809    +19013 13:00:22.069809

SQL> ALTER SESSION SET TIME_ZONE='00:00';

Session altered.

SQL> SELECT * from tmp_view;

       ZN                                 CZ                              C                                 TZ                              T                     DIFFZ                      DIFF
_________ __________________________________ ______________________________ __________________________________ ______________________________ _________________________ _________________________
+00:00    22/01/21 13:00:42,638457000 GMT    22/01/21 13:00:42,638457000    70/01/01 00:00:00,000000000 GMT    70/01/01 00:00:00,000000000    +19013 13:00:42.638457    +19013 13:00:42.638457
```

Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/oracle-appendonly/V8__add_user_managment.sql:30]: Currently with use `TIMESTAMP`s in a subquery like this:
```
SELECT sys_extract_utc(current_timestamp) - to_timestamp('1-1-1970 00:00:00','MM-DD-YYYY HH24:Mi:SS') as diff FROM dual
```

We could use `TIMESTAMP WITH TIME ZONE` if changed to this:
```
SELECT current_timestamp - to_timestamp_tz('1-1-1970 00:00:00 00:00','MM-DD-YYYY HH24:Mi:SS TZH:TZM') as diff FROM dual
```

The reason being docs are more explicit about arithmetic on `TIMESTAMP WITH TIME ZONE`

https://docs.oracle.com/cd/B19306_01/server.102/b14200/sql_elements001.htm#SQLRF00208:
> Oracle performs all timestamp arithmetic in UTC time. For TIMESTAMP WITH LOCAL TIME ZONE, Oracle converts the datetime value from the database time zone to UTC and converts back to the database time zone after performing the arithmetic. For TIMESTAMP WITH TIME ZONE, the datetime value is always in UTC, so no conversion is necessary.

Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/oracle-appendonly/V8__add_user_managment.sql:30]: fyi @meiersi-da @nmarton-da @rautenrieth-da

Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/oracle-appendonly/V8__add_user_managment.sql:30]: Excellent. Thank you @pbatko-da

**Review by mziolekda — APPROVED** [review]

**Comment by pbatko-da** [comment]
Closing this PR as we are moving away from DB side calculated timestamps in favor of calculating them on the JVM side.

**Pull request closed without merging**